arch/arm64/src/imx9/imx9_flexio_pwm.c

```diff
@@ -444,7 +444,7 @@ static int pwm_update_frequency(struct imx9_pwmtimer_s *priv, int freq)
 static int pwm_update_duty(struct imx9_pwmtimer_s *priv, int pwm_ch,
                            ub16_t duty16)
 {
-  uint64_t duty = ub16toi(duty16);
+  uint32_t duty = duty16 & 0xffffffff;
```

Comment:
I do not fully understand this change. Why you don't use `ub16toX` interface, but mask here?

Reply:
@pkarashchenko ub16toi() will discard the lower 16 bits, actually I'm more confused about the "ub16_t" type, it seems to be something NuttX specific. It is like an uint16_t but using higher 16 bits instead of lower 16 bits.

Reply:
I've been looking into SAMv7 PWM while in the past and need to refresh my knowledge. I will try to look today.

Reply:
SAMv7 uses `  width = b16toi(duty * period + b16HALF);`. Maybe we can use the same approach here?

Reply:
The equation is basically the same... ub16 is just fixed point value packed in 32 bits in form 16bits.16bits (16 bits integer + 16 bits decimal). The mask just takes the decimal part; and is not strictly needed as the input value is between 0-1 (upper 16 bits should be 0).

After multiplication, the result is then shifted right 16 bits (since there were 16 digits for decimals).

IMHO using ub macros makes it more confusing here; the ub16toui just does the shift, and obviously b16half is 0.5 in decimal and 0.1 in binary (0000.8000 in ub16 hex...)

This is the same form as what i used in the tpm_pwm driver

Reply:
Actually the interesting part here is that the PWM interface doesn't really define what is exactly the range of the input - it is between 0x0000.0000 and 0x0000ffff for sure, but what are the expected pulses at the limits? 

Many of the PWM HW blocks actually output one-clock pulses if duty is set to 0, and valid range is from 0x0 - 0x0000.ffff. And again, they produce one clock inactive cycle if the duty is set to maximum. That is, there is no real possibility to turn off the PWM by setting duty to 0 or to max, which would be intuitive.

And is there any REAL reason for the + 0.5 rounding ( + 0x0000.8000 or b16HALF), which is there in every PWM driver. +/-1 doesn't really make any difference in the 16-bit output (lack of rounding would produce an error which is just theoretical...) 

Anyhow, this implementation follows the same math as (most of the) other pwm drivers, and works in practice for us (running drone pwm ESCs / motors).

Reply:
@jlaitine I think @raiden00pl worked a lot of this low level issues and could have better understanding why to use or not these ub macros. In fact PWM range is uint16_t, that should be enough for most usage case, but not use if some MCU supports something like ultra resolution PWM on 32-bit.

Reply:
Personally I'm looking here from interface perspective. The `duty16` is `ub16_t` with the interface described in `fixedmath.h`. From the "back box" perspective the header provides a way to encode/decode (convert) data from/to integer and other definitions used for calculations, so I do not recommend to bypass that interface and try to do manual decoding. Line `uint32_t duty = duty16 & 0xffffffff;` does not makes any sense since `typedef uint32_t ub16_t;` and I just inspected that `uint32_t edge = (duty * priv->period + 0x8000) >> 16;` is exactly the same as `ub16toi(duty16 * priv->period  + b16HALF);` since `#define ub16toi(a)      ((a) >> 16)` and `#define b16HALF         0x00008000`. I would like to encourage you to use those interfaces as it leads toeasier review process.

Reply:
Ok, I will change it to the form which you like, no problem!

I will still explain why that form (or current way how ub16_t is used) doesn't make any sense.

1) normally when you use fixed point math, you need/want to make a decision about the range and resolution, i.e. select how many bits you use for interger and how many for decimals. This is trivial math (anyone who has done multiplication on pen and paper understands how the decimal point moves), and doesn't really need any "interface".

2. Allowing direct *, /, -, + operators on ub types is always wrong. If you consider e.g. the form which you suggest:

"ub16toi(duty16 * priv->period  + b16HALF);"

Here you must *know* that multiplying an ub16_t with an integer produces and ub16_t. Of course you know that after you have looked into internal definition of ub16_t. But wouldn't this look proper as well:

ub16_t a = uitou16_b(6);
ub 16_t b = uitou16_n(3);
ub16_t c = a * b; 
ub16_t d = a / b;

Both c and d are wrong. The multiplication returns ub32_t, which you again need to decipher by looking into the macros - c would result 0. The division returns integer. The d would result a small fraction (2/65536).

Or how about:
ub16_t a = uitoub16(5);
ub32_t b = uitoub32(2);
ub32_t c = a + b;

Again, looks proper, goes terribly wrong.

To make a proper interface, you'd need to define the types ub_x_t in a way that direct arithmetic is not allowed and use functions (ub16_mul, ub16_div, or in C++ by overloading operators).

Again in the same equation:

ub16toi(duty16 * priv->period  + b16HALF);

You are expected to *know* that the ub16toi doesnt round properly, but it just truncates. And that you need to round yourself. It would be clearer to have two functions ub16toi_floor and ub16toi_round. Or just the latter, I don't see why you wouldn't want to round the result properly.

This interface as such is just broken, that's why I dislike it and didn't want to touch it. It is much clearer to just note that the argument for pwm function is an uint32_t in range 0-65535, meaning pwm duty cycle between 0-100%.

I do respect the effort to try to abstact away the fixed point math. Just that fixed point math is so simple that I don't see the point to restrict yourself with this interface, and also if the interface itself is confusing and can be used in a wrong way. I don't also see how you would "review" those abominations; except by copy pasting the same form which is shown to work by magic earlier.

Btw. the interface *could* be improved. Just disallow the direct arithmetics by defining the ub16_t as structs containing just the single value. And define the proper (inline) functions to do the actual math.

Reply:
I agree that fixed math is tricky and I've been using 16.16 representations in the past as well as other amount of bits for fractional part and agree that having it configurable will be more flexible. Fixed math has a certain limitations and rules and obviously if you need to multiply of divide fixed math value it is not possible to use fixed math value for multiplier of divider. Maybe it is good to revise PWM interface in nuttx, so it can will be more straight forward, but definitely not is scope of current change, but should be done globally as a separate activity

Reply:
And just to mention; 16.16 representation is not tricky, nor is any other fixed point decimal representation. Enforcing usage of this interface makes it tricky. At least we should remove it's user side usage as it is not from any standard. In kernel internals we should either remove it or change it to enforce usage of functions instead of direct arithmetics (i.e. make it a real abstraction instead of current confusing "api")